File: pos-print/src/main/java/com/clouway/pos/print/service/GsonTransport.java
package com.clouway.pos.print.service;

import com.clouway.pos.print.client.JsonSerializer;
import com.clouway.pos.print.client.JsonSerializerFactory;
import com.google.inject.TypeLiteral;
import com.google.sitebricks.client.Transport;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * @author Miroslav Genov (mgenov@gmail.com)
 */
public final class GsonTransport implements Transport {

  private JsonSerializer serializer = JsonSerializerFactory.createSerializer();

  public <T> T in(InputStream inputStream, Class<T> tClass) throws IOException {
    return (T) serializer.deserializeEntity(inputStream, tClass);
  }

  @Override
  public <T> T in(InputStream in, TypeLiteral<T> type) throws IOException {
    return null;
  }

  public <T> void out(OutputStream outputStream, Class<T> tClass, T t) {
    String json = serializer.serializeEntity(t);
    try {
      outputStream.write(json.getBytes("UTF8"));
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  public String contentType() {
    return "application/json";
  }
}

File: pos-print/src/main/java/com/clouway/pos/print/transport/GsonTransport.kt
package com.clouway.pos.print.transport

import com.google.common.io.ByteStreams
import com.google.gson.GsonBuilder
import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import com.google.gson.stream.JsonWriter
import com.google.inject.Inject
import com.google.inject.TypeLiteral
import com.google.sitebricks.client.Transport
import java.io.*
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * @author Martin Milev <martin.milev@clouway.com>
 */
class GsonTransport @Inject
constructor() : Transport {

  class LocalDateTimeTypeAdapter : TypeAdapter<LocalDateTime>() {

    override fun write(out: JsonWriter, value: LocalDateTime?) {
      if (value == null) {
        out.nullValue()
        return
      }
      out.value(value.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME))
    }

    override fun read(reader: JsonReader): LocalDateTime? {
      if (reader.peek() == JsonToken.NULL) {
        reader.nextNull();
        return null;
      }

      val value = reader.nextString()
      return LocalDateTime.from(DateTimeFormatter.ISO_LOCAL_DATE_TIME.parse(value));
    }

  }

  internal val gson = GsonBuilder()
    .registerTypeAdapter(LocalDateTime::class.java, LocalDateTimeTypeAdapter())
    .create()
  

  @Throws(IOException::class)
  override fun <T> `in`(input: InputStream, tClass: Class<T>): T {
    return gson.fromJson(createReader(input), tClass)
  }

  @Throws(IOException::class)
  override fun <T> `in`(input: InputStream, typeLiteral: TypeLiteral<T>): T {
    return gson.fromJson<T>(createReader(input), typeLiteral.type)
  }

  @Throws(IOException::class)
  override fun <T> out(outputStream: OutputStream, tClass: Class<T>, t: T) {
    outputStream.write(gson.toJson(t).toByteArray(charset("UTF8")))
  }

  override fun contentType(): String {
    return "application/json"
  }

  @Throws(IOException::class)
  private fun createReader(input: InputStream): InputStreamReader {
    val out = ByteArrayOutputStream()
    ByteStreams.copy(input, out)
    return InputStreamReader(ByteArrayInputStream(out.toByteArray()), "UTF-8")
  }
}